### PR TITLE
feat: Added auth for generated openapi documents

### LINKED ISSUE #3183: Add auth to generated OpenAPI
The auth specified on an `Endpoint` should be reflected in the generated OpenAPI object

### PR DESCRIPTION
**Pull request by Gmin2**
The auth specified on an Endpoint is being reflected on the openapi components 

image

Fixes #3183 
/claim #3183

### CODE REVIEW

**Comment by Gmin2** [comment]
@987Nabil can you review this, idk why one test `ClientHttpsSpec.scala` is failing, it is non related to this part of work i had done in the PR

```
 - Https Client request - should throw DecoderException for handshake failure
Timeout of 1 m exceeded.
Executed in 4 m 17 s

[info] Completed tests
[info] Completed tests
[info] Completed tests
[error] Failed tests:
[error] 	zio.http.ClientHttpsSpec
[error] (zioHttpJVM / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 403 s (06:43), completed Dec 1, 2024, 7:14:39 AM

```

cc @jdegoes

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala [hunk 1]
```diff
@@ -223,6 +226,44 @@ object OpenAPIGenSpec extends ZIOSpecDefault {
...
+    Endpoint(GET / "withAuthQuery")
+      .in[SimpleInputBody]
+      .query(HttpCodec.query[String](apiKeyQueryParamName).optional)
+      .out[SimpleOutputBody]
+      .outError[NotFoundError](Status.NotFound)
+
+  private val basicAuthEndpoint =
+    Endpoint(GET / "withAuthHeader")
+      .in[SimpleInputBody]
+      .header(HttpCodec.authorization)
```
Comment (987Nabil) [inline zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:246]: The explicit mentioning of the auth header should not be needed. Please remove it. The tests have to be green without it.

**Review by 987Nabil — CHANGES_REQUESTED** [review]

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala [hunk 2]
```diff
@@ -223,6 +226,44 @@ object OpenAPIGenSpec extends ZIOSpecDefault {
...
+    Endpoint(GET / "withAuthHeader")
+      .in[SimpleInputBody]
+      .header(HttpCodec.Header.apply(apiKeyHeaderName, StringCodec).optional)
+      .out[SimpleOutputBody]
+      .outError[NotFoundError](Status.NotFound)
+
+  private val apiKeyAuthQueryEndpoint =
+    Endpoint(GET / "withAuthQuery")
+      .in[SimpleInputBody]
+      .query(HttpCodec.query[String](apiKeyQueryParamName).optional)
```
Comment (987Nabil) [inline zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:239]: This should be handled via `AuthType.Custom`

**Review by 987Nabil — CHANGES_REQUESTED** [review]

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 1 above)
Comment (Gmin2) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:246]: So we should use `.auth(AuthType)`

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 2 above)
Comment (Gmin2) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:239]: so this should be 
```scala
private val apiKeyAuthQueryEndpoint =
    Endpoint(GET / "withAuthQuery")
      .in[SimpleInputBody]
      .query(HttpCodec.query[String](apiKeyQueryParamName).optional)
      .out[SimpleOutputBody]
      .outError[NotFoundError](Status.NotFound)
      .auth(AuthType.Custom(HttpCodec.query[String](apiKeyQueryParamName).optional))
 ```
 
 cc @987Nabil ?

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 1 above)
Comment (Gmin2) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:246]: so `.header(HttpCodec.authorization)` should be removed from all this

```scala
private val bearerAuthEndpoint =
    Endpoint(GET / "withAuthHeader")
      .in[SimpleInputBody]
      .header(HttpCodec.authorization)
      .out[SimpleOutputBody]
      .outError[NotFoundError](Status.NotFound)
      .auth(AuthType.Bearer)

  private val digestAuthEndpoint =
    Endpoint(GET / "withAuthHeader")
      .in[SimpleInputBody]
      .header(HttpCodec.authorization)
      .out[SimpleOutputBody]
      .outError[NotFoundError](Status.NotFound)
      .auth(AuthType.Digest)  
  ```

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 2 above)
Comment (987Nabil) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:239]: Yes

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 1 above)
Comment (987Nabil) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:246]: Yes

**Comment by 987Nabil** [comment]
Closing due to inactivity
Feel free to reopen later on

**Pull request closed without merging**